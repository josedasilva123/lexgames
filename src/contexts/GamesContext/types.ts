import React from "react";
import { iGame } from "../@types/types";

export interface iGamesContext{
    addGame: (gameData: iGame) => void;
    removeGame: (clickedGame: iGame) => void;
    ratingGame: (gameId: number, rating: number) => void;
    filter: string;
    setFilter: React.Dispatch<React.SetStateAction<string>>;
    newFavoriteList: iGame[];
  }
  