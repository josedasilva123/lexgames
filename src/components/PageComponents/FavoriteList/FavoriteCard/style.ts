import styled from "styled-components";

//Estilos menos dinâmicos também podem ser comtemplados no 

export const StyledFavoriteCard = styled.li`
    display: flex;
    align-items: center;
    justify-content: space-between;    
    gap: 2rem;
    padding: 16px;
    border: 1px solid var(--color-lightblue);
    border-radius: 4px;

    .favoriteRating{
        display: flex;
        align-items: center;
        gap: .6rem;
        p{
            font-weight: 600;
        }
    }

    @media (max-width: 450px){
        align-items: flex-start;
        flex-direction: column;

        button{
            width: 100%;   
        }
    } 
`