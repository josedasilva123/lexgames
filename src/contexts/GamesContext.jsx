/* eslint-disable no-restricted-globals */
import { createContext, useContext } from "react";
import { coreApi } from "../services/api";
import { UserContext } from "./UserContext";
import { toast } from "react-toastify";
import { userActions } from "./UserContext";

export const GamesContext = createContext({});

export const GamesProvider = ({ children }) => {
  const { user, dispatchUser } = useContext(UserContext);

  const addGame = async (gameData) => {
    if (!user?.favoriteGames.find((game) => game.id === gameData.id)) {
      try {
        const token = localStorage.getItem("@TOKEN");

        const newFavoriteList = [
          ...user?.favoriteGames,
          {
            id: gameData.id,
            title: gameData.title,
            thumbnail: gameData.thumbnail,
            genre: gameData.genre,
            rating: null,
          },
        ];

        await coreApi.patch(
          "user/favorites",
          {
            favoriteGames: newFavoriteList,
          },
          {
            headers: {
              auth: token,
            },
          }
        ); /* Atualiza o banco de dados */

        toast.success("Jogo adicionado com sucesso!");
        dispatchUser({ type: userActions.setFavoriteList, payload: newFavoriteList });
      } catch (error) {
        console.log(error);
        toast.error(error.response.data.error);
      }
    } else {
      toast.alert("Este jogo já está cadastrado na lista.");
    }
  };

  const removeGame = async (clickedGame) => {
    if (confirm("Deseja excluir mesmo este item?")) {
      try {
        const newFavoriteList = user?.favoriteGames.filter((game) => game.id !== clickedGame.id);
        const token = localStorage.getItem("@TOKEN");
         
        await coreApi.patch(
          "user/favorites",
          {
            favoriteGames: newFavoriteList,
          },
          {
            headers: {
              auth: token,
            },
          }
        );

        toast.success("Jogo removido com sucesso!");
        dispatchUser({ type: userActions.setFavoriteList, payload: newFavoriteList });
      } catch (error) {
        console.log(error);
        toast.error(error.response.data.error);
      }
    }

    if (user?.favoriteGames.length === 1) {
      localStorage.removeItem("@FAVORITE_LIST");
    }
  };

  const ratingGame = async (gameId, rating) => {
    try {
      const token = localStorage.getItem("@TOKEN");
      
      const newFavoriteList = user?.favoriteGames.map((game) => {
        if (game.id === gameId) {
          return { ...game, rating: rating };
        } else {
          return game;
        }
      });

      await coreApi.patch(
        "user/favorites",
        {
          favoriteGames: newFavoriteList,
        },
        {
          headers: {
            auth: token,
          },
        }
      );

      toast.success("Avaliação realizada com sucesso!");
      dispatchUser({ type: userActions.setFavoriteList, payload: newFavoriteList });
    } catch (error) {
      console.log(error);
      toast.error(error.response.data.error);
    }   
  };

  return (
    <GamesContext.Provider value={{ addGame, removeGame, ratingGame }}>
      {children}
    </GamesContext.Provider>
  );
};
