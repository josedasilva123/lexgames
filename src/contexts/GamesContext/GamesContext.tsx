/* eslint-disable no-restricted-globals */
import { createContext, useContext } from "react";
import { coreApi } from "../../services/api";
import { UserContext } from "../UserContext";
import { toast } from "react-toastify";
import { iApiError, iDefaultContextProps, iGame } from "../types/types";
import { AxiosError } from "axios";
import { iGamesContext } from "./types";

export const GamesContext = createContext({} as iGamesContext);

export const GamesProvider = ({ children }: iDefaultContextProps) => {
  const { favoriteList, setFavoriteList } = useContext(UserContext);

  const addGame = async (gameData: iGame) => {
    if (!favoriteList.find((game) => game.id === gameData.id)) {
      try {
        const token = localStorage.getItem("@TOKEN");

        const newData = [
          ...favoriteList,
          {
            id: gameData.id,
            title: gameData.title,
            thumbnail: gameData.thumbnail,
            genre: gameData.genre,
            rating: null,
          },
        ];

        await coreApi.patch(
          "user/favorites",
          {
            favoriteGames: newData,
          },
          {
            headers: {
              auth: token as string,
            },
          }
        ); /* Atualiza o banco de dados */

        toast.success("Jogo adicionado com sucesso!");
        setFavoriteList(newData); /* Estado */
      } catch (error) {
        const requestError = error as AxiosError<iApiError>
        toast.error(requestError.response?.data.error);
      }
    } else {
      toast.warn("Este jogo já está cadastrado na lista.");
    }
  };

  const removeGame = async (clickedGame: iGame) => {
    if (confirm("Deseja excluir mesmo este item?")) {
      try {
        const newGameList = favoriteList.filter((game) => game.id !== clickedGame.id);
        const token = localStorage.getItem("@TOKEN");

        await coreApi.patch(
          "user/favorites",
          {
            favoriteGames: newGameList,
          },
          {
            headers: {
              auth: token as string,
            },
          }
        );

        toast.success("Jogo removido com sucesso!");
        setFavoriteList(newGameList);
      } catch (error) {
        const requestError = error as AxiosError<iApiError>
        toast.error(requestError.response?.data.error);
      }
    }

    if (favoriteList.length === 1) {
      localStorage.removeItem("@FAVORITE_LIST");
    }
  };

  const ratingGame = async (gameId: number, rating: number) => {
    try {
      const token = localStorage.getItem("@TOKEN");
      
      const newFavoriteList = favoriteList.map((game) => {
        if (game.id === gameId) {
          return { ...game, rating: rating };
        } else {
          return game;
        }
      });

      await coreApi.patch(
        "user/favorites",
        {
          favoriteGames: newFavoriteList,
        },
        {
          headers: {
            auth: token as string,
          },
        }
      );

      toast.success("Avaliação realizada com sucesso!")
      setFavoriteList(newFavoriteList);
    } catch (error) {
      const requestError = error as AxiosError<iApiError>
      toast.error(requestError.response?.data.error);
    }   
  };

  return (
    <GamesContext.Provider value={{ addGame, removeGame, ratingGame }}>
      {children}
    </GamesContext.Provider>
  );
};
