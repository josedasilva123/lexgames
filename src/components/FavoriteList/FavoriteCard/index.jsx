import React, { useContext } from "react";
import { GamesContext } from "../../../contexts/GamesContext";
import { StyledButton } from "../../../styles/button";
import { StyledParagraph, StyledTitle } from "../../../styles/typography";
import StarInput from "../../CustomInputs/StarInput";
import { StyledFavoriteCard } from "./style";

const FavoriteCard = ({ game }) => {
  const { removeGame, ratingGame } = useContext(GamesContext);
  return (
    <StyledFavoriteCard>
      <div>
        <StarInput currentRating={game.rating ? game.rating : 0} callback={(number) => ratingGame(game.id, number)} />
        {game.rating && <StyledParagraph>{game.rating}</StyledParagraph>}
        <StyledParagraph>{game.genre}</StyledParagraph>
       
        <StyledTitle tag="h3" fontSize="two">
          {game.title}
        </StyledTitle>
      </div>
      <StyledButton buttonStyle="solid2" onClick={() => removeGame(game)}>Remover</StyledButton>
    </StyledFavoriteCard>
  );
};

export default FavoriteCard;
