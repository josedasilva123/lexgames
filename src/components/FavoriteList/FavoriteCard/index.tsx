import React, { useContext } from "react";
import { GamesContext } from "../../../contexts/GamesContext/GamesContext";
import { iGame } from "../../../contexts/types/types";
import { StyledButton } from "../../../styles/button";
import { StyledParagraph, StyledTitle } from "../../../styles/typography";
import StarInput from "../../CustomInputs/StarInput";
import { StyledFavoriteCard } from "./style";

interface iFavoriteCardProps{
  game: iGame;
}

const FavoriteCard = ({ game }: iFavoriteCardProps) => {
  const { removeGame, ratingGame } = useContext(GamesContext);
  return (
    <StyledFavoriteCard>
      <div>
        <StarInput currentRating={game.rating ? game.rating : 0} callback={(number: number) => ratingGame(game.id, number)} />
        {game.rating && <StyledParagraph>{game.rating}</StyledParagraph>}
        <StyledParagraph>{game.genre}</StyledParagraph>
       
        <StyledTitle tag="h3" fontSize="two">
          {game.title}
        </StyledTitle>
      </div>
      <StyledButton buttonStyle="solid2" onClick={() => removeGame(game)}>Remover</StyledButton>
    </StyledFavoriteCard>
  );
};

export default FavoriteCard;
