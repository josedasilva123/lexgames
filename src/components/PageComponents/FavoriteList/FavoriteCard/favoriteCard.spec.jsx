/* eslint-disable testing-library/await-async-query */
import { screen, render, fireEvent, waitFor } from "@testing-library/react";
import FavoriteCard from ".";
import { GamesContext } from "../../../../contexts/GamesContext/GamesContext";

const mockGame = {
  id: 1,
  title: "League of Legends",
  thumbnail: "https://example.com/image.png",
  genre: "MOBA",
  rating: 5,
};

const removeGame = jest.fn();

describe("<FavoriteCard />", () => {
  it("should contain all the game information", async () => {
    render(<FavoriteCard game={mockGame} />);
    const category = screen.getByText("MOBA");
    const title = screen.getByText("League of Legends");
    const rating = screen.getByText("5");

    expect(category).toBeInTheDocument();
    expect(title).toBeInTheDocument();
    expect(rating).toBeInTheDocument();
  });

  it("should fire event on click in remove button", async () => {
    render(
      <GamesContext.Provider value={{ removeGame }}>
        <FavoriteCard game={mockGame} />
      </GamesContext.Provider>
    );

    const removeButton = screen.getByText("Remover");

    fireEvent.click(removeButton);

    await waitFor(() => {
      expect(removeGame).toHaveBeenCalledTimes(1);
    });
  });

  it("should match with snapshot", () => {
    const { container } = render(<FavoriteCard game={mockGame} />);
    expect(container).toMatchSnapshot();
  });
});
