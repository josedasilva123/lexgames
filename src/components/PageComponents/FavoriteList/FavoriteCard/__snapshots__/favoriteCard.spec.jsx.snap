// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteCard /> should match with snapshot 1`] = `
<div>
  <li
    class="sc-gKXOVf dlcKCP"
  >
    <div>
      <div
        class="favoriteRating"
      >
        <p
          class="sc-hKMtZM hZQxYS"
        >
          5
        </p>
        <fieldset
          class="sc-jSMfEi bzfCno"
        >
          <span
            role="listitem"
          >
            <svg
              fill="yellow"
              height="21"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            role="listitem"
          >
            <svg
              fill="yellow"
              height="21"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            role="listitem"
          >
            <svg
              fill="yellow"
              height="21"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            role="listitem"
          >
            <svg
              fill="yellow"
              height="21"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            role="listitem"
          >
            <svg
              fill="yellow"
              height="21"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="21"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </fieldset>
      </div>
      <p
        class="sc-hKMtZM hZQxYS"
      >
        MOBA
      </p>
      <h3
        class="sc-dkzDqf UfQEd"
      >
        League of Legends
      </h3>
    </div>
    <button
      class="sc-bczRLJ xIDPU"
    >
      Remover
    </button>
  </li>
</div>
`;
